docs/language/ql-handbook/types.rst

```diff
@@ -479,6 +480,47 @@ program, so it's helpful to extend a new type (namely ``TTaintType``)::
     class Tainted extends TaintType, TTaintedValue {
     }
     
+.. _type-unions:
+
+Type unions
+***********
+
+.. note::
+    The syntax for type unions is considered experimental and is subject to change.
+    However, type unions appear in the `standard QL libraries <https://github.com/github/codeql>`__.
+    The following sections should help you understand those examples.
+
+Type unions are user-defined types that are declared with the keyword ``class``.
+The syntax resembles :ref:`type aliases <type-aliases>`, but with two or more type expressions on the right-hand side.
+
+Type unions are used for creating restricted versions of an existing :ref:`algebraic datatype <algebraic-datatypes>`, by explicitly
+selecting a subset of the branches of that datatype and binding them to a new type.
+Type unions of :ref:`database types <database-types>` are also supported.
+
+You can use a type union to explicitly restrict the permitted branches from an algebraic datatype
+and resolve spurious :ref:`recursion <recursion>` in predicates.
+For example, the following construction is legal::
+
+    newtype T =
+      T1(T t) { not exists(T2orT3 s | t = s) } or
+      T2(int x) { x = 1 or x = 2 } or
+      T3(int x) { x = 3 or x = 4 or x = 5 }
+
+    class T2orT3 = T2 or T3;
```

Comment:
This example looks slightly wonky. If I understand the semantics correctly, it seems that the `T1` branch would end up being empty, because every possible binding for `t` will be rejected by `not exists(T2orT3 s | t = s)`.

Can we find a less vacuous example? I can't off the top of my head come up with something that doesn't need an auxiliary predicate, but perhaps something like
```ql
newtype T =
    T1(int x) { x in [ 1..9] and not exists(T2orT3 s | x = extract(s) } or
    T2(int x) { x = 1 or x = 2 } or
    T3(int x) { x = 3 or x = 4 or x = 5 }

class T2orT3 = T2 or T3;

int extract(T2orT3 t) { t = T2(result) or t = T3(result) }
```

Reply:
True, that example comes for one of the PR tests and is pretty vacuous. Yours looks better.
It would probably be even nicer to have something more inspired by real code, but I don't have a good insight into that. Perhaps @dbartol would know what to do?

Reply:
Hmm, how about something like (for at least a smell of program analysis)
```ql
newtype InitialValueSource =
    ExplicitInitialization(VarDecl v) { exists(v.getInitializer()) } or
    ParameterPassing(Call c, int pos) { exists(c.getParameter(pos)) } or
    UnknownInitialGarbage(VarDecl v) { not exists(DefiniteInitialization di | v = target(di)) }

class DefiniteInitialization = ParameterPassing or ExplicitInitialization;

VarDecl target(DefiniteInitialization di) {
   di = ExplicitInitialization(result) or
   exists(Call c, int pos | di = ParameterPassing(c, pos) and
                            result = c.getCallee().getFormalArg(pos))
}
```

Reply:
That looks great, thank you very much for taking the time to come up with this!